### ISSUE TITLE
error getting certificate for sub-subdomain

### ISSUE THREAD

**Issue by emfrias** [caddy-dns/cloudflare#13]
When I try to get a certificate for an address like a.b.c.com instead of just b.c.com, it fails.  I'm using v2.2.0-rc.1.  Not sure if this is a bug or if I'm missing something in the configuration.

Caddyfile:
```
foo.bar.baz.com {
    tls me@baz.com {
        dns cloudflare my-api-key-goes-here
        ca https://acme-staging-v02.api.letsencrypt.org/directory
    }
}
```
Gives the error:
```
2020/08/12 23:33:19 [INFO] [foo.bar.baz.com] acme: use dns-01 solver
2020/08/12 23:33:19 [INFO] [foo.bar.baz.com] acme: Preparing to solve DNS-01
2020/08/12 23:33:20 [INFO] [foo.bar.baz.com] acme: Cleaning DNS-01 challenge
2020/08/12 23:33:20 [WARN] [foo.bar.baz.com] acme: cleaning up failed: no memory of presenting a DNS record for foo.bar.baz.com 
2020/08/12 23:33:20 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/93420171
2020/08/12 23:33:20 [ERROR] error: one or more domains had a problem:
[foo.bar.baz.com] [foo.bar.baz.com] acme: error presenting token: got error status: HTTP 403: [{Code:0 Message:Actor 'com.cloudflare.api.token.c8a1af1c44011cc50326adf273f7413c' requires permission 'com.cloudflare.api.account.zone.read' to list zones}]
 (challenge=dns-01 remaining=[])
2020/08/12 23:33:22 [ERROR] attempt 1: [foo.bar.baz.com] Obtain: [foo.bar.baz.com] error: one or more domains had a problem:
[foo.bar.baz.com] [foo.bar.baz.com] acme: error presenting token: got error status: HTTP 403: [{Code:0 Message:Actor 'com.cloudflare.api.token.c8a1af1c44011cc50326adf273f7413c' requires permission 'com.cloudflare.api.account.zone.read' to list zones}]
 - retrying in 1m0s (3.988028658s/720h0m0s elapsed)...
```

The API key I'm using has the required Zone.Zone and Zone.DNS permissions, and everything works as it should if I change the hostname to `bar.baz.com` in the Caddyfile.  I can also use this same API key with another ACME client ([acme.sh](https://github.com/acmesh-official/acme.sh)) and sucessfully get a certificate for `foo.bar.baz.com`.

On a side note, Caddy v1 also had problems with this config, it would print a message like:
```
[foo.bar.baz.com] [foo.bar.baz.com] acme: error presenting token: cloudflare: failed to find zone bar.baz.com.: Zone could not be found
```

p.s, thanks for Caddy, I haven't used v2 much yet, but v1 has been wonderfully easy to configure and stable.

**Comment by mholt** [comment]
Make sure you're using an API token, not an API key: https://github.com/libdns/cloudflare#authenticating

If `c8a1af1c44011cc50326adf273f7413c` is in fact your API token, then you've just posted it publicly, FYI. You will need to change it.

I'm not sure how to fix this otherwise, since frankly, it works for me... you'll just need to make sure the credentials are configured properly.

**Comment by emfrias** [comment]
Thanks for the response.  Your assurance that it did work forced me to go back and do some more testing, and it turns out the problem is very specific to my setup, and it's quite possible there's nothing wrong with Caddy.  

I'm doing split-horizon DNS on bar.baz.com (really, the only reason I'm using DNS challenges is because I'm trying to get certificates for hostnames that only resolve on my local network but letsencrypt can't see them.  On my local network, I have two nameservers, one for bar.baz.com, and another for baz.com which delegates all queries for *.bar.baz.com to the first server.  It seems like there's something about this setup that's causing the problem, because I can successfully get certificates for foo.anything-except-bar.baz.com.

I found that if I change my DNS resolver to use 8.8.8.8 while getting the certificate, it works correctly (I'm running Caddy in docker, so I pass `--dns 8.8.8.8`)

I don't understand it yet, but my current theory is that Caddy tries to figure out what zone foo.bar.baz.com belongs to, and it does that by asking the local DNS servers "is bar.baz.com a zone?".  Then, if not, asking "is baz.com a zone?"  For my local network, bar.baz.com is a zone, and it assumes that will be the case on CloudFlare too.  Apparently, if I had an Enterprise plan, CloudFlare would allow me to create a "child subdomain" which would function this way.  I only have a free account which doesn't give me this option.

Is there any other logging I can turn on to figure out if I'm on the right track?


p.s., You're right, I used the wrong term in my original issue, I was using "API Tokens" configured according to your link, they were not "API Keys".  And I had sanitized the logs I posted, but thanks for looking out--next time I'll redact the sensitive stuff in an obvious way instead of replacing it with similar-looking random numbers.

**Comment by mholt** [comment]
> I'm doing split-horizon DNS on baz.com

In Caddy 2.2, you can configure the DNS resolver to use for the DNS challenge (this is also how lego works behind split-horizon DNS): https://github.com/caddyserver/caddy/commit/744d04c2585d50f64cf3d43d139c310a18e78f73

Caddyfile support for that should be coming soon.

> I don't understand it yet, but my current theory is that Caddy tries to figure out what zone foo.bar.baz.com belongs to, and it does that by asking the local DNS servers "is bar.baz.com a zone?". Then, if not, asking "is baz.com a zone?"

Correct!

> Is there any other logging I can turn on to figure out if I'm on the right track?

Shouldn't be necessary at this point, since you can now configure a custom DNS resolver to bypass your LAN's tricks. Please give it a try if you can, would be nice to verify it is working before tagging the release!

**Comment by emfrias** [comment]
Thanks, that should solve my problem.  I'm having trouble working out the right Caddyfile syntax though:

```bash
$ xcaddy build 24f34780b61fc82df6507cf95535c3dfc9f78058 --with github.com/caddy-dns/cloudflare --with github.com/caddyserver/jsonc-adapter
$ ./caddy version
v2.2.0-rc.1.0.20200831192326-24f34780b61f h1:sof4puBUKjIjEdXSTDIJaRF87TQg4MAjjTh+May/GOU=
```
I guessed
```
foo.bar.baz.com {
    tls me@baz.com {
        dns cloudflare my-api-key-goes-here
        ca https://acme-staging-v02.api.letsencrypt.org/directory
        resolvers 8.8.8.8
    }
}
```
but that wouldn't parse.  Tried a few other places too, but no luck.

I then tried the JSON syntax with:
```json
{
  "apps": {
    "http": {
      "servers": {
        "srv0": {
          "listen": [
            ":443"
          ],
          "routes": [
            {
              "match": [
                {
                  "host": [
                    "foo.bar.baz.com"
                  ]
                }
              ],
              "terminal": true
            }
          ]
        }
      }
    },
    "tls": {
      "automation": {
        "policies": [
          {
            "subjects": [
              "foo.bar.baz.com"
            ],
            "issuer": {
              "ca": "https://acme-staging-v02.api.letsencrypt.org/directory",
              "challenges": {
                "dns": {
                  "provider": {
                    "api_token": "---redacted---",
                    "name": "cloudflare"
                  },
                  "resolvers": [
                    "8.8.8.8",
                    "8.8.4.4"
                  ]
                }
              },
              "email": "me@baz.com",
              "module": "acme"
            }
          }
        ]
      }
    }
  }
}
```

It runs with this (and I think I'm putting the config in the right place, because if I misspell "resolvers" it will error out).  But unfortunately I still get the same error (when "adding temporary record for zone bar.baz.com").  Do you see anything I'm doing wrong?

I tried my trick of running the same caddy & config file in a docker container with "--dns 8.8.8.8" and it still worked (no surprise there)

**Comment by mholt** [comment]
Hmm, yeah, that config looks right... it should be prioritizing your choice there and then fall back to the system resolvers and finally some hard-coded ones as a last resort.

Ohhh wait, did you try with `:53` at the end of each address? You might have to manually specify the port. I didn't think about that.

Relevant code here: https://github.com/caddyserver/certmagic/blob/09acc6bf59fafc220fd444667851de44e55c1419/solvers.go#L290

See the `defaultNameservers` variable.

**Comment by emfrias** [comment]
Same result with `"resolvers": ["8.8.8.8:53", "8.8.4.4:53"]`, and with just one: `"resolvers": ["8.8.8.8:53"]`

**Comment by mholt** [comment]
Okay now I'm not sure why it's not working. I can circle back on this later, but for a quicker resolution, if you have some time to debug that would be very helpful while I work through my backlog! ✌️

**Comment by emfrias** [comment]
I'll give it a shot, you've given me some good clues where to look.  I'm not familiar with go's package system and build system so I have a bit of stumbling around ahead of me.

**Comment by mholt** [comment]
Is this still not working?

**Comment by will528** [comment]
I just wanted to add that I'm also experiencing this issue. I haven't found a Caddyfile syntax that correctly sets up a custom DNS resolver for acme verification.

As a workaround, I've just set my whole server system to use Cloudflare's 1.1.1.1 as it's DNS server instead of my network's server with the split records. Not ideal, but it works for now.

**Comment by mholt** [comment]
@will528 Would you mind digging into the code and finding out why? I haven't found anyone who can reproduce the issue and trace it down. Add some `fmt.Println()` statements to see what is going on.

**Comment by will528** [comment]
@mholt I'm glad you pointed me to the code. I'm still learning Go but the syntax made a million times more sense when I read through [acmeissuer.go](https://github.com/caddyserver/caddy/blob/master/modules/caddytls/acmeissuer.go). User-error on my part.

I didn't realize that the dns subdirective is a shortcut to an acme template ([explained here](https://caddyserver.com/docs/caddyfile/directives/tls)). This Caddyfile worked to produce the correct JSON configuration, I just had to manually specify the acme parameters:

```
example.com {
    root * /var/caddy/example.com
    file_server
    tls {
        issuer acme {
            dir https://acme-staging-v02.api.letsencrypt.org/directory
            dns cloudflare <api_token>
            resolvers 1.1.1.1 1.0.0.1
        }
    }
}
```

This JSON is produced:
```json
{
    "apps": {
      "http": {...},
      "tls": {
        "automation": {
          "policies": [
            {
              "subjects": ["example.com"],
              "issuers": [
                {
                  "ca": "https://acme-staging-v02.api.letsencrypt.org/directory",
                  "challenges": {
                    "dns": {
                      "provider": {
                        "api_token": "<api_token>",
                        "name": "cloudflare"
                      },
                      "resolvers": ["1.1.1.1", "1.0.0.1"]
                    }
                  },
                  "module": "acme"
                }
              ]
            }
          ]
        }
      }
    }
  }
```

Log output when DNS is using my internal DNS:
```
2020/11/21 20:39:54.266	ERROR	tls.obtain	will retry	{"error": "[example.com] Obtain: [example.com] solving challenges: presenting for challenge: adding temporary record for zone com.: expected 1 zone, got 0 for com. (order=https://acme-staging-v02.api.letsencrypt.org/acme/order/.../...) (ca=https://acme-staging-v02.api.letsencrypt.org/directory)", "attempt": 1, "retrying_in": 60, "elapsed": 0.886847867, "max_duration": 2592000}
```

Log output using above Caddyfile, successfully hitting 1.1.1.1 for the challenge:
```
2020/11/21 20:41:56.102	INFO	tls.issuance.acme.acme_client	trying to solve challenge	{"identifier": "example.com", "challenge_type": "dns-01", "ca": "https://acme-staging-v02.api.letsencrypt.org/directory"}
2020/11/21 20:41:59.483	INFO	tls.issuance.acme.acme_client	validations succeeded; finalizing order	{"order": "https://acme-staging-v02.api.letsencrypt.org/acme/order/.../..."}
2020/11/21 20:41:59.851	INFO	tls.issuance.acme.acme_client	successfully downloaded available certificate chains	{"count": 2, "first_url": "https://acme-staging-v02.api.letsencrypt.org/acme/cert/..."}
2020/11/21 20:41:59.852	INFO	tls.obtain	certificate obtained successfully	{"identifier": "example.com"}
```

**Comment by mholt** [comment]
@will528 Awesome, glad you figured out how the code works! So, uh, it's working for you with your Caddyfile then? So this is resolved?

**Comment by will528** [comment]
I think so! I'm curious if @emfrias would have their issue resolved now on 2.2.1. They mentioned they were experiencing the issue on 2.2.0, and the related commits I see are tagged with 2.2.1. I'm on 2.2.1 and it's working perfectly with the Caddyfile I included above.

Really appreciate the help!

**Comment by mholt** [comment]
Great! I'll close the issue then but if it's confirmed to still be an issue, we can reopen it... I hope not though.
